projects/train/train/waveform_sampler.py

```diff
@@ -0,0 +1,254 @@
+            Number of samples to load from each file at a time.
+            This is useful for reducing I/O overhead when reading.
+        path:
+            Optional path to location of datasets in hdf5 files.
+            `path` should be delimited by forward slashes. If `None`
+            it is assumed the datasets are at the root of the file.
+    """
+
+    def __init__(
+        self,
+        fnames: Iterable[Path],
+        channels: Iterable[str],
+        batch_size: int,
+        batches_per_epoch: int,
+        chunk_size: int = 1000,
+        path: Optional[Path] = None,
+    ):
+        self.fnames = fnames
+        self.channels = channels
+        self.batch_size = batch_size
+        self.batches_per_epoch = batches_per_epoch
+        self.chunk_size = chunk_size
+
+        if path is not None:
+            self.path = path.split("/")
+        else:
+            self.path = None
+
+        self.sizes = {}
+        self.mmap_files = {}
+        self.mmap_datasets = {}
+
+        # for each file store the datasets
+        # of interest in a dictionary so we
+        # can access them at will without needing
+        # to reopen the files each time
+        for fname in self.fnames:
+            f, g = self.open(fname)
+            self.mmap_files[fname] = f
+            self.mmap_datasets[fname] = {
+                channel: g[channel] for channel in self.channels
+            }
+
+            # store sizes of each dataset and warn if not chunked;
+            # assumes all dsets have same attributes
+            # like size and chunking behavior
+            dset = self.mmap_datasets[fname][self.channels[0]]
+            self.sizes[fname] = len(dset)
+            if dset.chunks is None:
+                warnings.warn(
+                    "File {} contains datasets that were generated "
+                    "without using chunked storage. This can have "
+                    "severe performance impacts at data loading time. "
+                    "If you need faster loading, try re-generating "
+                    "your datset with chunked storage turned on.".format(
+                        fnames
+                    ),
+                )
+
+        self.waveform_size = dset.shape[1]
+        self.probs = np.array([i / self.total for i in self.sizes.values()])
+
+    @property
+    def num_channels(self):
+        return len(self.channels)
+
+    @property
+    def chunks_per_batch(self):
+        return math.ceil(self.batch_size / self.chunk_size)
+
+    @property
+    def total(self):
+        return sum(self.sizes.values())
+
+    def __len__(self):
+        return self.batches_per_epoch
+
+    def __del__(self):
+        # close all opened files when the object is destroyed
+        for f in self.mmap_files.values():
+            f.close()
+
+    def open(self, fname) -> tuple[h5py.File, h5py.Group]:
+        f = group = h5py.File(fname, "r")
+        if self.path is not None:
+            for path in self.path:
```

Comment:
I think you could just use the full string, including the `/`, to index into the file. E.g., `f["a/b/c"]` should be the same as `f["a"]["b"]["c"]`